src/software/ai/hl/stp/play/free_kick/free_kick_play_fsm.cpp

```diff
@@ -0,0 +1,322 @@
+    const std::optional<Point> &pass_origin_override)
+{
+    // These two tactics will set robots to roam around the field, trying to put
+    // themselves into a good position to receive a pass
+    if (num_tactics != receiver_positioning_tactics.size())
+    {
+        receiver_positioning_tactics =
+            std::vector<std::shared_ptr<MoveTactic>>(num_tactics);
+        std::generate(receiver_positioning_tactics.begin(),
+                      receiver_positioning_tactics.end(),
+                      []() { return std::make_shared<MoveTactic>(); });
+    }
+
+    std::vector<Point> best_receiving_positions =
+        receiver_position_generator.getBestReceivingPositions(
+            *world, num_tactics, existing_receiver_positions, pass_origin_override);
+    // Note that getBestReceivingPositions may return fewer positions than requested
+    // if there are not enough robots, so we will need to check the size of the vector.
+    for (unsigned int i = 0;
+         i < receiver_positioning_tactics.size() && i < best_receiving_positions.size();
+         i++)
+    {
+        Angle receiver_orientation =
+            (world->ball().position() - best_receiving_positions[i]).orientation();
+        receiver_positioning_tactics[i]->updateControlParams(
+            best_receiving_positions[i], receiver_orientation, 0.0,
+            TbotsProto::MaxAllowedSpeedMode::PHYSICAL_LIMIT,
+            TbotsProto::ObstacleAvoidanceMode::AGGRESSIVE);
+    }
+}
+
+void FreeKickPlayFSM::setReceiverAndDefenderTactics(
+    PriorityTacticVector &tactics_to_run, const Update &event, int num_receivers,
+    int num_defenders, const std::vector<Point> &existing_receiver_positions,
+    const std::optional<Point> &pass_origin_override)
+{
+    if (num_receivers > 0)
+    {
+        updateReceiverPositioningTactics(event.common.world_ptr, num_receivers,
+                                         existing_receiver_positions,
+                                         pass_origin_override);
+        tactics_to_run[0].insert(tactics_to_run[0].end(),
+                                 receiver_positioning_tactics.begin(),
+                                 receiver_positioning_tactics.end());
+    }
+
+    defense_play->updateControlParams(TbotsProto::MaxAllowedSpeedMode::PHYSICAL_LIMIT);
+
+    if (num_defenders > 0)
+    {
+        defense_play->updateTactics(PlayUpdate(
+            event.common.world_ptr, num_defenders,
+            [&tactics_to_run](PriorityTacticVector new_tactics) {
+                for (const auto &tactic_vector : new_tactics)
+                {
+                    tactics_to_run.push_back(tactic_vector);
+                }
+            },
+            event.common.inter_play_communication,
+            event.common.set_inter_play_communication_fun));
+    }
+}
+
+void FreeKickPlayFSM::updateAlignToBallTactic(const WorldPtr &world_ptr)
+{
+    // Face towards the center of the field
+    Point position_to_face(0, 0);
```

Comment:
why face the centre of the field? why not face the enemy goal centre?

Reply:
I was thinking that most of the time we will be passing it to a friendly robot that is near the center of the field. We could face the enemy goal center, but I was thinking that it would be rare for the net to be open enough (goalie out of the way, or crease defenders not blocking the net properly) for us to actually take a direct shot on it. Though in the rare case where the net is open enough, we would just pivot around the ball and take the shot anyways as part of the `ShootState_S`.

Reply:
Actually on second thought, I was just watching the fourth test case in the pytest, and the center of the field may not always be a great decision, specially if we're around the center of the enemy half (robot aligns with ball facing backwards). I decided to update this to be aiming towards the center of the left segment of the enemy defense area instead. This should be good in scenarios where the ball is in the friendly half or enemy half. The main edge case I think this is better than the enemy goal center is when we're taking a corner kick.